In [1]:
import pandas as pd

Having multi-layer index allows you to more effectively categorize data

In [17]:
# This module uses the Big Mac Index dataset
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'])
bigmac.head()

Date    Country  Price in US Dollars
0 2016-01-01  Argentina                 2.39
1 2016-01-01  Australia                 3.74
2 2016-01-01     Brazil                 3.35
3 2016-01-01    Britain                 4.22
4 2016-01-01     Canada                 4.14

In [8]:
bigmac.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 3 columns):
Date                   652 non-null datetime64[ns]
Country                652 non-null object
Price in US Dollars    652 non-null float64
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 15.4+ KB


# Section 7; Part 95
Create a multiindex with the `set_index()` method

In [9]:
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'])

In [10]:
# Can pass multiple values to `set_index()`
bigmac.set_index(keys=["Date", "Country"])

Price in US Dollars
Date       Country                            
2016-01-01 Argentina                      2.39
           Australia                      3.74
           Brazil                         3.35
           Britain                        4.22
           Canada                         4.14
           Chile                          2.94
           China                          2.68
           Colombia                       2.43
           Costa Rica                     4.02
           Czech Republic                 2.98
           Denmark                        4.32
           Egypt                          2.16
           Euro area                      4.00
           Hong Kong                      2.48
           Hungary                        3.08
           India                          1.90
           Indonesia                      2.19
           Israel                         4.29
           Japan                          3.12
           Malaysia                       1.82
           Mexico                         2.81
           New Zealand                    3.91
           Norway                         5.21
           Pakistan                       2.86
           Peru                           2.93
           Philippines                    2.79
           Poland                         2.37
           Russia                         1.53
           Saudi Arabia                   3.20
           Singapore                      3.27
...                                        ...
2010-01-01 Hong Kong                      1.91
           Hungary                        3.86
           Indonesia                      2.24
           Israel                         3.99
           Japan                          3.50
           Latvia                         3.09
           Lithuania                      2.87
           Malaysia                       2.08
           Mexico                         2.50
           New Zealand                    3.61
           Norway                         7.02
           Pakistan                       2.42
           Peru                           2.81
           Philippines                    2.21
           Poland                         2.86
           Russia                         2.34
           Saudi Arabia                   2.67
           Singapore                      3.19
           South Africa                   2.46
           South Korea                    2.98
           Sri Lanka                      1.83
           Sweden                         5.51
           Switzerland                    6.30
           Taiwan                         2.36
           Thailand                       2.11
           Turkey                         3.83
           UAE                            2.99
           Ukraine                        1.83
           United States                  3.58
           Uruguay                        3.32

[652 rows x 1 columns]

In [12]:
# Can change how the indexes are structured by changing order of passed index columns
bigmac.set_index(keys = ['Country', 'Date'])

Price in US Dollars
Country        Date                           
Argentina      2016-01-01                 2.39
Australia      2016-01-01                 3.74
Brazil         2016-01-01                 3.35
Britain        2016-01-01                 4.22
Canada         2016-01-01                 4.14
Chile          2016-01-01                 2.94
China          2016-01-01                 2.68
Colombia       2016-01-01                 2.43
Costa Rica     2016-01-01                 4.02
Czech Republic 2016-01-01                 2.98
Denmark        2016-01-01                 4.32
Egypt          2016-01-01                 2.16
Euro area      2016-01-01                 4.00
Hong Kong      2016-01-01                 2.48
Hungary        2016-01-01                 3.08
India          2016-01-01                 1.90
Indonesia      2016-01-01                 2.19
Israel         2016-01-01                 4.29
Japan          2016-01-01                 3.12
Malaysia       2016-01-01                 1.82
Mexico         2016-01-01                 2.81
New Zealand    2016-01-01                 3.91
Norway         2016-01-01                 5.21
Pakistan       2016-01-01                 2.86
Peru           2016-01-01                 2.93
Philippines    2016-01-01                 2.79
Poland         2016-01-01                 2.37
Russia         2016-01-01                 1.53
Saudi Arabia   2016-01-01                 3.20
Singapore      2016-01-01                 3.27
...                                        ...
Hong Kong      2010-01-01                 1.91
Hungary        2010-01-01                 3.86
Indonesia      2010-01-01                 2.24
Israel         2010-01-01                 3.99
Japan          2010-01-01                 3.50
Latvia         2010-01-01                 3.09
Lithuania      2010-01-01                 2.87
Malaysia       2010-01-01                 2.08
Mexico         2010-01-01                 2.50
New Zealand    2010-01-01                 3.61
Norway         2010-01-01                 7.02
Pakistan       2010-01-01                 2.42
Peru           2010-01-01                 2.81
Philippines    2010-01-01                 2.21
Poland         2010-01-01                 2.86
Russia         2010-01-01                 2.34
Saudi Arabia   2010-01-01                 2.67
Singapore      2010-01-01                 3.19
South Africa   2010-01-01                 2.46
South Korea    2010-01-01                 2.98
Sri Lanka      2010-01-01                 1.83
Sweden         2010-01-01                 5.51
Switzerland    2010-01-01                 6.30
Taiwan         2010-01-01                 2.36
Thailand       2010-01-01                 2.11
Turkey         2010-01-01                 3.83
UAE            2010-01-01                 2.99
Ukraine        2010-01-01                 1.83
United States  2010-01-01                 3.58
Uruguay        2010-01-01                 3.32

[652 rows x 1 columns]

Common practice is to have the outer layers of the index have *fewer* values than the inner layers. This would be the first example above

In [18]:
bigmac.set_index(keys=["Date", "Country"], inplace=True)
bigmac.head()

Price in US Dollars
Date       Country                       
2016-01-01 Argentina                 2.39
           Australia                 3.74
           Brazil                    3.35
           Britain                   4.22
           Canada                    4.14

In [21]:
# Sorting the index, will sort all indexes
bigmac.sort_index(inplace=True)
bigmac.head()

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76
           Britain                   3.67
           Canada                    3.97

In [22]:
# Index attribute will return a `MultiIndex` object representing each level of the index
bigmac.index

MultiIndex(levels=[[2010-01-01 00:00:00, 2010-07-01 00:00:00, 2011-07-01 00:00:00, 2012-01-01 00:00:00, 2012-07-01 00:00:00, 2013-01-01 00:00:00, 2013-07-01 00:00:00, 2014-01-01 00:00:00, 2014-07-01 00:00:00, 2015-01-01 00:00:00, 2015-07-01 00:00:00, 2016-01-01 00:00:00], [u'Argentina', u'Australia', u'Austria', u'Belgium', u'Brazil', u'Britain', u'Canada', u'Chile', u'China', u'Colombia', u'Costa Rica', u'Czech Republic', u'Denmark', u'Egypt', u'Estonia', u'Euro area', u'Finland', u'France', u'Germany', u'Greece', u'Hong Kong', u'Hungary', u'India', u'Indonesia', u'Ireland', u'Israel', u'Italy', u'Japan', u'Latvia', u'Lithuania', u'Malaysia', u'Mexico', u'Netherlands', u'New Zealand', u'Norway', u'Pakistan', u'Peru', u'Philippines', u'Poland', u'Portugal', u'Russia', u'Saudi Arabia', u'Singapore', u'South Africa', u'South Korea', u'Spain', u'Sri Lanka', u'Sweden', u'Switzerland', u'Taiwan', u'Thailand', u'Turkey', u'UAE', u'Ukraine', u'United States', u'Uruguay', u'Venezuela', u'Vietn

In [24]:
# Get the index names
bigmac.index.names

FrozenList([u'Date', u'Country'])

In [25]:
# MultiIndexes are a different object type from single index dataframes 
type(bigmac.index)

pandas.indexes.multi.MultiIndex

In [26]:
# Indexes are a tuple of all levels of the index
# Because of this, we'll need to pass a tuple to get specific cell values in the next steps
bigmac.index[0]

(Timestamp('2010-01-01 00:00:00'), 'Argentina')

# Section 7; Part 96
The `get_level_values()` method

In [28]:
# Set index while importing and sort to improve speed
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'], index_col = ['Date', 'Country'])
bigmac.sort_index(inplace=True)
bigmac.head()

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76
           Britain                   3.67
           Canada                    3.97

In [29]:
# get_level_values() is called on the index
# Accepts either a numeric value or a name
# This function does not de-duplicate a layer
bigmac.index.get_level_values('Date')

DatetimeIndex(['2010-01-01', '2010-01-01', '2010-01-01', '2010-01-01',
               '2010-01-01', '2010-01-01', '2010-01-01', '2010-01-01',
               '2010-01-01', '2010-01-01',
               ...
               '2016-01-01', '2016-01-01', '2016-01-01', '2016-01-01',
               '2016-01-01', '2016-01-01', '2016-01-01', '2016-01-01',
               '2016-01-01', '2016-01-01'],
              dtype='datetime64[ns]', name=u'Date', length=652, freq=None)

In [30]:
# Pass a numeric value (0 = Date, 1 = Country)
bigmac.index.get_level_values(1)

Index([u'Argentina', u'Australia', u'Brazil', u'Britain', u'Canada', u'Chile',
       u'China', u'Colombia', u'Costa Rica', u'Czech Republic',
       ...
       u'Switzerland', u'Taiwan', u'Thailand', u'Turkey', u'UAE', u'Ukraine',
       u'United States', u'Uruguay', u'Venezuela', u'Vietnam'],
      dtype='object', name=u'Country', length=652)

# Section 7; Part 97
The `set_names()` method

 - Like `get_level_values()`, this is called on the `.index` object

In [31]:
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'], index_col = ['Date', 'Country'])
bigmac.sort_index(inplace=True)
bigmac.head()

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76
           Britain                   3.67
           Canada                    3.97

In [35]:
# Change "Date" to "Day" and "Country" to "Location"
bigmac.index.set_names(['Day', 'Location'], inplace=True)
bigmac.head()

Price in US Dollars
Day        Location                      
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76
           Britain                   3.67
           Canada                    3.97

# Section 7, Part 98
The `sort_index()` method

In [36]:
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'], index_col = ['Date', 'Country'])

In [38]:
# By default, it will sort all indexes in ascending order
bigmac.sort_index()

Price in US Dollars
Date       Country                            
2010-01-01 Argentina                      1.84
           Australia                      3.98
           Brazil                         4.76
           Britain                        3.67
           Canada                         3.97
           Chile                          3.18
           China                          1.83
           Colombia                       3.91
           Costa Rica                     3.52
           Czech Republic                 3.71
           Denmark                        5.99
           Egypt                          2.38
           Euro area                      4.84
           Hong Kong                      1.91
           Hungary                        3.86
           Indonesia                      2.24
           Israel                         3.99
           Japan                          3.50
           Latvia                         3.09
           Lithuania                      2.87
           Malaysia                       2.08
           Mexico                         2.50
           New Zealand                    3.61
           Norway                         7.02
           Pakistan                       2.42
           Peru                           2.81
           Philippines                    2.21
           Poland                         2.86
           Russia                         2.34
           Saudi Arabia                   2.67
...                                        ...
2016-01-01 Italy                          4.30
           Japan                          3.12
           Malaysia                       1.82
           Mexico                         2.81
           Netherlands                    3.71
           New Zealand                    3.91
           Norway                         5.21
           Pakistan                       2.86
           Peru                           2.93
           Philippines                    2.79
           Poland                         2.37
           Portugal                       3.23
           Russia                         1.53
           Saudi Arabia                   3.20
           Singapore                      3.27
           South Africa                   1.77
           South Korea                    3.59
           Spain                          3.76
           Sri Lanka                      2.43
           Sweden                         5.23
           Switzerland                    6.44
           Taiwan                         2.08
           Thailand                       3.09
           Turkey                         3.41
           UAE                            3.54
           Ukraine                        1.54
           United States                  4.93
           Uruguay                        3.74
           Venezuela                      0.66
           Vietnam                        2.67

[652 rows x 1 columns]

In [40]:
# Alter sort to date in ascending order by county in descending by passing a list to `ascending` parameter
bigmac.sort_index(ascending = [True, False], inplace=True)
bigmac.head()

Price in US Dollars
Date       Country                           
2010-01-01 Uruguay                       3.32
           United States                 3.58
           Ukraine                       1.83
           UAE                           2.99
           Turkey                        3.83

# Section 7; Part 99
Extracting rows from a multiindex dataframe

 - Done using `loc[]` and `ix[]` 

In [41]:
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'], index_col = ['Date', 'Country'])
bigmac.sort_index(inplace=True)

In [42]:
# Pass a tuple of the combination of indexes we are looking for
# If we put 1 layer, it pulls everything for that index
bigmac.loc[('2010-01-01')]

Price in US Dollars
Date       Country                            
2010-01-01 Argentina                      1.84
           Australia                      3.98
           Brazil                         4.76
           Britain                        3.67
           Canada                         3.97
           Chile                          3.18
           China                          1.83
           Colombia                       3.91
           Costa Rica                     3.52
           Czech Republic                 3.71
           Denmark                        5.99
           Egypt                          2.38
           Euro area                      4.84
           Hong Kong                      1.91
           Hungary                        3.86
           Indonesia                      2.24
           Israel                         3.99
           Japan                          3.50
           Latvia                         3.09
           Lithuania                      2.87
           Malaysia                       2.08
           Mexico                         2.50
           New Zealand                    3.61
           Norway                         7.02
           Pakistan                       2.42
           Peru                           2.81
           Philippines                    2.21
           Poland                         2.86
           Russia                         2.34
           Saudi Arabia                   2.67
           Singapore                      3.19
           South Africa                   2.46
           South Korea                    2.98
           Sri Lanka                      1.83
           Sweden                         5.51
           Switzerland                    6.30
           Taiwan                         2.36
           Thailand                       2.11
           Turkey                         3.83
           UAE                            2.99
           Ukraine                        1.83
           United States                  3.58
           Uruguay                        3.32

In [43]:
# Extract something specific - Date/Country combination needed
bigmac.loc[('2010-01-01', 'Brazil')]

Price in US Dollars    4.76
Name: (2010-01-01 00:00:00, Brazil), dtype: float64

In [46]:
# Extract something specific - Date/Country combination needed - and pull only 1 column instead of entire series
bigmac.loc[('2010-01-01', 'Brazil'), "Price in US Dollars"]

Date        Country
2010-01-01  Brazil     4.76
Name: Price in US Dollars, dtype: float64

In [47]:
# Extract 2015-07-01
bigmac.loc['2015-07-01']

Price in US Dollars
Date       Country                            
2015-07-01 Argentina                      3.07
           Australia                      3.92
           Austria                        3.71
           Belgium                        4.05
           Brazil                         4.28
           Britain                        4.51
           Canada                         4.54
           Chile                          3.27
           China                          2.74
           Colombia                       2.92
           Costa Rica                     4.03
           Czech Republic                 2.83
           Denmark                        5.08
           Egypt                          2.16
           Estonia                        3.23
           Euro area                      4.05
           Finland                        4.49
           France                         4.49
           Germany                        3.93
           Greece                         3.34
           Hong Kong                      2.48
           Hungary                        3.18
           India                          1.83
           Indonesia                      2.29
           Ireland                        4.05
           Israel                         4.63
           Italy                          4.38
           Japan                          2.99
           Malaysia                       2.01
           Mexico                         3.11
           Netherlands                    3.78
           New Zealand                    3.91
           Norway                         5.65
           Pakistan                       3.44
           Peru                           3.14
           Philippines                    3.61
           Poland                         2.54
           Portugal                       3.29
           Russia                         1.88
           Saudi Arabia                   3.20
           Singapore                      3.44
           South Africa                   2.09
           South Korea                    3.76
           Spain                          4.00
           Sri Lanka                      2.61
           Sweden                         5.13
           Switzerland                    6.82
           Taiwan                         2.55
           Thailand                       3.17
           Turkey                         3.87
           UAE                            3.54
           Ukraine                        1.55
           United States                  4.79
           Uruguay                        4.13
           Venezuela                      0.67
           Vietnam                        2.75

In [48]:
# Using the `ix` method, doesn't show outer layer - Notice "Date" is gone
bigmac.ix[('2016-01-01')]

Price in US Dollars
Country                            
Argentina                      2.39
Australia                      3.74
Austria                        3.76
Belgium                        4.25
Brazil                         3.35
Britain                        4.22
Canada                         4.14
Chile                          2.94
China                          2.68
Colombia                       2.43
Costa Rica                     4.02
Czech Republic                 2.98
Denmark                        4.32
Egypt                          2.16
Estonia                        3.23
Euro area                      4.00
Finland                        4.41
France                         4.41
Germany                        3.86
Greece                         3.60
Hong Kong                      2.48
Hungary                        3.08
India                          1.90
Indonesia                      2.19
Ireland                        4.25
Israel                         4.29
Italy                          4.30
Japan                          3.12
Malaysia                       1.82
Mexico                         2.81
Netherlands                    3.71
New Zealand                    3.91
Norway                         5.21
Pakistan                       2.86
Peru                           2.93
Philippines                    2.79
Poland                         2.37
Portugal                       3.23
Russia                         1.53
Saudi Arabia                   3.20
Singapore                      3.27
South Africa                   1.77
South Korea                    3.59
Spain                          3.76
Sri Lanka                      2.43
Sweden                         5.23
Switzerland                    6.44
Taiwan                         2.08
Thailand                       3.09
Turkey                         3.41
UAE                            3.54
Ukraine                        1.54
United States                  4.93
Uruguay                        3.74
Venezuela                      0.66
Vietnam                        2.67

In [50]:
# Return a series
bigmac.ix[('2016-01-01', 'Brazil')]

Price in US Dollars    3.35
Name: (2016-01-01 00:00:00, Brazil), dtype: float64

In [51]:
# Get specific value instead of entire series
bigmac.ix[('2016-01-01', 'Brazil'), "Price in US Dollars"]

3.3500000000000001

# Section 7; Part 100
The `transpose()` method

 - `transpose()` - Swaps axises
 - Columns can be multindexed as well

In [52]:
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'], index_col = ['Date', 'Country'])
bigmac.sort_index(inplace=True)

In [54]:
# Doesn't have an inplace, so it needs to be reassigned
bigmac = bigmac.transpose()
bigmac.head()

Date                2010-01-01                                              \
Country              Argentina Australia Brazil Britain Canada Chile China   
Price in US Dollars       1.84      3.98   4.76    3.67   3.97  3.18  1.83   

Date                                                     ...    2016-01-01  \
Country             Colombia Costa Rica Czech Republic   ...   Switzerland   
Price in US Dollars     3.91       3.52           3.71   ...          6.44   

Date                                                                    \
Country             Taiwan Thailand Turkey   UAE Ukraine United States   
Price in US Dollars   2.08     3.09   3.41  3.54    1.54          4.93   

Date                                           
Country             Uruguay Venezuela Vietnam  
Price in US Dollars    3.74      0.66    2.67  

[1 rows x 652 columns]

In [55]:
# Extract 
#  Now the index is only single level - Price in US Dollars
# This will return a multiple index series
bigmac.ix["Price in US Dollars"]

Date        Country       
2010-01-01  Argentina         1.84
            Australia         3.98
            Brazil            4.76
            Britain           3.67
            Canada            3.97
            Chile             3.18
            China             1.83
            Colombia          3.91
            Costa Rica        3.52
            Czech Republic    3.71
            Denmark           5.99
            Egypt             2.38
            Euro area         4.84
            Hong Kong         1.91
            Hungary           3.86
            Indonesia         2.24
            Israel            3.99
            Japan             3.50
            Latvia            3.09
            Lithuania         2.87
            Malaysia          2.08
            Mexico            2.50
            New Zealand       3.61
            Norway            7.02
            Pakistan          2.42
            Peru              2.81
            Philippines       2.21
            Poland          

In [56]:
# Extract specific value by passing an index to the second parameter
bigmac.ix['Price in US Dollars', ('2010-01-01','Brazil')]

4.7599999999999998

In [57]:
# Return a series for date
bigmac.ix['Price in US Dollars', ('2010-01-01')]

Country
Argentina         1.84
Australia         3.98
Brazil            4.76
Britain           3.67
Canada            3.97
Chile             3.18
China             1.83
Colombia          3.91
Costa Rica        3.52
Czech Republic    3.71
Denmark           5.99
Egypt             2.38
Euro area         4.84
Hong Kong         1.91
Hungary           3.86
Indonesia         2.24
Israel            3.99
Japan             3.50
Latvia            3.09
Lithuania         2.87
Malaysia          2.08
Mexico            2.50
New Zealand       3.61
Norway            7.02
Pakistan          2.42
Peru              2.81
Philippines       2.21
Poland            2.86
Russia            2.34
Saudi Arabia      2.67
Singapore         3.19
South Africa      2.46
South Korea       2.98
Sri Lanka         1.83
Sweden            5.51
Switzerland       6.30
Taiwan            2.36
Thailand          2.11
Turkey            3.83
UAE               2.99
Ukraine           1.83
United States     3.58
Uruguay           3.32
Nam

MultiIndex can be on both Axis, to extract on these you'd have two tuples for each parameter in `ix` or `loc`

# Section 7; Part 101
The `swaplevel()` method

 - Swaps levels of multiindex
 - If there are only two indexes, no parameters are needed

In [59]:
bigmac = pd.read_csv('datasets/bigmac.csv', parse_dates=['Date'], index_col = ['Date', 'Country'])
bigmac.sort_index(inplace=True)
bigmac.head()

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76
           Britain                   3.67
           Canada                    3.97

In [62]:
# No inplace parameter, so it needs to be reassigned
bigmac = bigmac.swaplevel()
bigmac.head()

Price in US Dollars
Country   Date                           
Argentina 2010-01-01                 1.84
Australia 2010-01-01                 3.98
Brazil    2010-01-01                 4.76
Britain   2010-01-01                 3.67
Canada    2010-01-01                 3.97

# Section 7; Part 102
The `stack()` method

 - Takes the columns axis and moves them to the horizontal axis

In [65]:
world = pd.read_csv('datasets/worldstats.csv', index_col = ["country", "year"])
world.head()

Population           GDP
country    year                           
Arab World 2015  392022276.0  2.530102e+12
           2014  384222592.0  2.873600e+12
           2013  376504253.0  2.846994e+12
           2012  368802611.0  2.773270e+12
           2011  361031820.0  2.497945e+12

In [66]:
world.stack()

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
                  GDP           2.873600e+12
            2013  Population    3.765043e+08
                  GDP           2.846994e+12
            2012  Population    3.688026e+08
                  GDP           2.773270e+12
            2011  Population    3.610318e+08
                  GDP           2.497945e+12
            2010  Population    3.531122e+08
                  GDP           2.103825e+12
            2009  Population    3.450542e+08
                  GDP           1.798878e+12
            2008  Population    3.368865e+08
                  GDP           2.081343e+12
            2007  Population    3.287666e+08
                  GDP           1.641666e+12
            2006  Population    3.209067e+08
                  GDP           1.404190e+12
            2005  Population    3.134309e+08
                  GDP     

Since our dataframe was converted to one column, it became a Series

In [67]:
# Convert the above series to a dataframe instead of to a series
# Notice we have a column with no name so it was given "0"
world.stack().to_frame()

0
country    year                         
Arab World 2015 Population  3.920223e+08
                GDP         2.530102e+12
           2014 Population  3.842226e+08
                GDP         2.873600e+12
           2013 Population  3.765043e+08
                GDP         2.846994e+12
           2012 Population  3.688026e+08
                GDP         2.773270e+12
           2011 Population  3.610318e+08
                GDP         2.497945e+12
           2010 Population  3.531122e+08
                GDP         2.103825e+12
           2009 Population  3.450542e+08
                GDP         1.798878e+12
           2008 Population  3.368865e+08
                GDP         2.081343e+12
           2007 Population  3.287666e+08
                GDP         1.641666e+12
           2006 Population  3.209067e+08
                GDP         1.404190e+12
           2005 Population  3.134309e+08
                GDP         1.184735e+12
           2004 Population  3.063874e+08
                GDP         9.639180e+11
           2003 Population  2.997320e+08
                GDP         8.218702e+11
           2002 Population  2.934026e+08
                GDP         7.279939e+11
           2001 Population  2.872918e+08
                GDP         7.221506e+11
...                                  ...
Zimbabwe   1974 Population  5.967861e+06
                GDP         3.982161e+09
           1973 Population  5.768382e+06
                GDP         3.309354e+09
           1972 Population  5.573312e+06
                GDP         2.677729e+09
           1971 Population  5.385342e+06
                GDP         2.178716e+09
           1970 Population  5.206311e+06
                GDP         1.884206e+09
           1969 Population  5.036321e+06
                GDP         1.747999e+09
           1968 Population  4.874113e+06
                GDP         1.479600e+09
           1967 Population  4.718612e+06
                GDP         1.397002e+09
           1966 Population  4.568320e+06
                GDP         1.281750e+09
           1965 Population  4.422132e+06
                GDP         1.311436e+09
           1964 Population  4.279561e+06
                GDP         1.217138e+09
           1963 Population  4.140804e+06
                GDP         1.159512e+09
           1962 Population  4.006262e+06
                GDP         1.117602e+09
           1961 Population  3.876638e+06
                GDP         1.096647e+09
           1960 Population  3.752390e+06
                GDP         1.052990e+09

[22422 rows x 1 columns]

# Section 7; Part 103
`unstack()` (Part I)

 - Does the reverse of `stack()` starting with the inner most layer by default

In [69]:
world = pd.read_csv('datasets/worldstats.csv', index_col = ["country", "year"])
s = world.stack()
s.head()

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
                  GDP           2.873600e+12
            2013  Population    3.765043e+08
dtype: float64

In [70]:
# Undo the stack
s.unstack()

Population           GDP
country     year                          
Afghanistan 1960   8994793.0  5.377778e+08
            1961   9164945.0  5.488889e+08
            1962   9343772.0  5.466667e+08
            1963   9531555.0  7.511112e+08
            1964   9728645.0  8.000000e+08
            1965   9935358.0  1.006667e+09
            1966  10148841.0  1.400000e+09
            1967  10368600.0  1.673333e+09
            1968  10599790.0  1.373333e+09
            1969  10849510.0  1.408889e+09
            1970  11121097.0  1.748887e+09
            1971  11412821.0  1.831109e+09
            1972  11716896.0  1.595555e+09
            1973  12022514.0  1.733333e+09
            1974  12315553.0  2.155555e+09
            1975  12582954.0  2.366667e+09
            1976  12831361.0  2.555556e+09
            1977  13056499.0  2.953333e+09
            1978  13222547.0  3.300000e+09
            1979  13283279.0  3.697940e+09
            1980  13211412.0  3.641723e+09
            1981  12996923.0  3.478788e+09
            2001  20531160.0  2.461666e+09
            2002  21487079.0  4.128818e+09
            2003  22507368.0  4.583649e+09
            2004  23499850.0  5.285462e+09
            2005  24399948.0  6.275076e+09
            2006  25183615.0  7.057598e+09
            2007  25877544.0  9.843842e+09
            2008  26528741.0  1.019053e+10
...                      ...           ...
Zimbabwe    1986   9198874.0  6.217524e+09
            1987   9535657.0  6.741215e+09
            1988   9866776.0  7.814784e+09
            1989  10184966.0  8.286323e+09
            1990  10484771.0  8.783817e+09
            1991  10763036.0  8.641482e+09
            1992  11019717.0  6.751472e+09
            1993  11256512.0  6.563813e+09
            1994  11476807.0  6.890675e+09
            1995  11683136.0  7.111271e+09
            1996  11877664.0  8.553147e+09
            1997  12059858.0  8.529572e+09
            1998  12226742.0  6.401968e+09
            1999  12374019.0  6.858013e+09
            2000  12499981.0  6.689958e+09
            2001  12603988.0  6.777385e+09
            2002  12691431.0  6.342116e+09
            2003  12774162.0  5.727592e+09
            2004  12867828.0  5.805598e+09
            2005  12984418.0  5.755215e+09
            2006  13127942.0  5.443896e+09
            2007  13297798.0  5.291950e+09
            2008  13495462.0  4.415703e+09
            2009  13720997.0  8.157077e+09
            2010  13973897.0  9.422161e+09
            2011  14255592.0  1.095623e+10
            2012  14565482.0  1.239272e+10
            2013  14898092.0  1.349023e+10
            2014  15245855.0  1.419691e+10
            2015  15602751.0  1.389294e+10

[11211 rows x 2 columns]

In [71]:
# Can keep unstacking
#  This will create a multiindex column
#  In this case having a Population with year subgroups and GDP with year subgroups
s.unstack().unstack()

Population                              \
year                            1960          1961          1962   
country                                                            
Afghanistan             8.994793e+06  9.164945e+06  9.343772e+06   
Albania                          NaN           NaN           NaN   
Algeria                 1.112489e+07  1.140486e+07  1.169015e+07   
Andorra                          NaN           NaN           NaN   
Angola                           NaN           NaN           NaN   
Antigua and Barbuda              NaN           NaN           NaN   
Arab World                       NaN           NaN           NaN   
Argentina                        NaN           NaN  2.128768e+07   
Armenia                          NaN           NaN           NaN   
Aruba                            NaN           NaN           NaN   
Australia               1.027648e+07  1.048300e+07  1.074200e+07   
Austria                 7.047539e+06  7.086299e+06  7.129864e+06   
Azerbaijan                       NaN           NaN           NaN   
Bahamas, The            1.095260e+05  1.151080e+05  1.210830e+05   
Bahrain                          NaN           NaN           NaN   
Bangladesh              4.820070e+07  4.959361e+07  5.103060e+07   
Barbados                         NaN           NaN           NaN   
Belarus                          NaN           NaN           NaN   
Belgium                 9.153489e+06  9.183948e+06  9.220578e+06   
Belize                  9.206800e+04  9.470100e+04  9.738900e+04   
Benin                   2.431620e+06  2.466002e+06  2.503232e+06   
Bermuda                 4.440000e+04  4.550000e+04  4.660000e+04   
Bhutan                           NaN           NaN           NaN   
Bolivia                 3.693451e+06  3.764815e+06  3.838096e+06   
Bosnia and Herzegovina           NaN           NaN           NaN   
Botswana                5.240290e+05  5.365760e+05  5.499900e+05   
Brazil                  7.249358e+07  7.470689e+07  7.700755e+07   
Brunei Darussalam                NaN           NaN           NaN   
Bulgaria                         NaN           NaN           NaN   
Burkina Faso            4.829291e+06  4.894578e+06  4.960325e+06   
...                              ...           ...           ...   
Switzerland             5.327827e+06  5.434294e+06  5.573815e+06   
Syrian Arab Republic    4.592777e+06  4.742596e+06  4.897428e+06   
Tajikistan                       NaN           NaN           NaN   
Tanzania                         NaN           NaN           NaN   
Thailand                2.739718e+07  2.822421e+07  2.908104e+07   
Timor-Leste                      NaN           NaN           NaN   
Togo                    1.580513e+06  1.597528e+06  1.612758e+06   
Tonga                            NaN           NaN           NaN   
Trinidad and Tobago     8.484810e+05  8.653560e+05  8.800190e+05   
Tunisia                          NaN           NaN           NaN   
Turkey                  2.755328e+07  2.822929e+07  2.890998e+07   
Turkmenistan                     NaN           NaN           NaN   
Tuvalu                           NaN           NaN           NaN   
Uganda                  6.788211e+06  7.006629e+06  7.240155e+06   
Ukraine                          NaN           NaN           NaN   
United Arab Emirates             NaN           NaN           NaN   
United Kingdom          5.240000e+07  5.280000e+07  5.325000e+07   
United States           1.806710e+08  1.836910e+08  1.865380e+08   
Upper middle income     1.154592e+09  1.159145e+09  1.176093e+09   
Uruguay                 2.538651e+06  2.571691e+06  2.603887e+06   
Uzbekistan                       NaN           NaN           NaN   
Vanuatu                          NaN           NaN           NaN   
Venezuela, RB           8.146845e+06  8.461684e+06  8.790590e+06   
Vietnam                          NaN           NaN           NaN   
Virgin Islands (U.S.)   3.200000e+04  3.410000e+04  3.630000e+04   
West Bank and 

In [72]:
# Can unstack once more
s.unstack().unstack().unstack()

            year  country               
Population  1960  Afghanistan               8.994793e+06
                  Albania                            NaN
                  Algeria                   1.112489e+07
                  Andorra                            NaN
                  Angola                             NaN
                  Antigua and Barbuda                NaN
                  Arab World                         NaN
                  Argentina                          NaN
                  Armenia                            NaN
                  Aruba                              NaN
                  Australia                 1.027648e+07
                  Austria                   7.047539e+06
                  Azerbaijan                         NaN
                  Bahamas, The              1.095260e+05
                  Bahrain                            NaN
                  Bangladesh                4.820070e+07
                  Barbados                     

# Section 7; Part 104
`unstack()` (Part II)

In [73]:
world = pd.read_csv('datasets/worldstats.csv', index_col = ["country", "year"])

In [75]:
s = world.stack()
s.head()

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
                  GDP           2.873600e+12
            2013  Population    3.765043e+08
dtype: float64

In [79]:
# unstack
#  Provide an index (name or numeric index) that we want to unstack
#   Country =0; Year =1; Unnamed (Pop/GDP) = 2
s.unstack("year")

year                                      1960          1961          1962  \
country                                                                      
Afghanistan           Population  8.994793e+06  9.164945e+06  9.343772e+06   
                      GDP         5.377778e+08  5.488889e+08  5.466667e+08   
Albania               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Algeria               Population  1.112489e+07  1.140486e+07  1.169015e+07   
                      GDP         2.723638e+09  2.434767e+09  2.001461e+09   
Andorra               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Angola                Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Antigua and Barbuda   Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Arab World            Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Argentina             Population           NaN           NaN  2.128768e+07   
                      GDP                  NaN           NaN  2.445060e+10   
Armenia               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Aruba                 Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Australia             Population  1.027648e+07  1.048300e+07  1.074200e+07   
                      GDP         1.856759e+10  1.963938e+10  1.988353e+10   
Austria               Population  7.047539e+06  7.086299e+06  7.129864e+06   
                      GDP         6.592694e+09  7.311750e+09  7.756110e+09   
Azerbaijan            Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Bahamas, The          Population  1.095260e+05  1.151080e+05  1.210830e+05   
                      GDP         1.698023e+08  1.900962e+08  2.122528e+08   
Bahrain               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
...                                        ...           ...           ...   
United Arab Emirates  Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
United Kingdom        Population  5.240000e+07  5.280000e+07  5.325000e+07   
                      GDP         7.232805e+10  7.669436e+10  8.060194e+10   
United States         Population  1.806710e+08  1.836910e+08  1.865380e+08   
                      GDP         5.433000e+11  5.633000e+11  6.051000e+11   
Upper middle income   Population  1.154592e+09  1.159145e+09  1.176093e+09   
                      GDP         1.854612e+11  1.690732e+11  1.777957e+11   
Uruguay               Population  2.538651e+06  2.571691e+06  2.603887e+06   
                      GDP         1.242289e+09  1.547389e+09  1.710004e+09   
Uzbekistan            Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Vanuatu               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Venezuela, RB         Population  8.146845e+06  8.461684e+06  8.790590e+06   
                      GDP         8.607600e+09  8.923367e+09  9.873398e+09   
Vietnam               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Virgin Islands (U.S.) 

In [81]:
# Providing negative arguments will start at inner most level and work from there
#  This will do the same as above (because year is the second from the inner most)
s.unstack(-2)

year                                      1960          1961          1962  \
country                                                                      
Afghanistan           Population  8.994793e+06  9.164945e+06  9.343772e+06   
                      GDP         5.377778e+08  5.488889e+08  5.466667e+08   
Albania               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Algeria               Population  1.112489e+07  1.140486e+07  1.169015e+07   
                      GDP         2.723638e+09  2.434767e+09  2.001461e+09   
Andorra               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Angola                Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Antigua and Barbuda   Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Arab World            Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Argentina             Population           NaN           NaN  2.128768e+07   
                      GDP                  NaN           NaN  2.445060e+10   
Armenia               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Aruba                 Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Australia             Population  1.027648e+07  1.048300e+07  1.074200e+07   
                      GDP         1.856759e+10  1.963938e+10  1.988353e+10   
Austria               Population  7.047539e+06  7.086299e+06  7.129864e+06   
                      GDP         6.592694e+09  7.311750e+09  7.756110e+09   
Azerbaijan            Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Bahamas, The          Population  1.095260e+05  1.151080e+05  1.210830e+05   
                      GDP         1.698023e+08  1.900962e+08  2.122528e+08   
Bahrain               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
...                                        ...           ...           ...   
United Arab Emirates  Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
United Kingdom        Population  5.240000e+07  5.280000e+07  5.325000e+07   
                      GDP         7.232805e+10  7.669436e+10  8.060194e+10   
United States         Population  1.806710e+08  1.836910e+08  1.865380e+08   
                      GDP         5.433000e+11  5.633000e+11  6.051000e+11   
Upper middle income   Population  1.154592e+09  1.159145e+09  1.176093e+09   
                      GDP         1.854612e+11  1.690732e+11  1.777957e+11   
Uruguay               Population  2.538651e+06  2.571691e+06  2.603887e+06   
                      GDP         1.242289e+09  1.547389e+09  1.710004e+09   
Uzbekistan            Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Vanuatu               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Venezuela, RB         Population  8.146845e+06  8.461684e+06  8.790590e+06   
                      GDP         8.607600e+09  8.923367e+09  9.873398e+09   
Vietnam               Population           NaN           NaN           NaN   
                      GDP                  NaN           NaN           NaN   
Virgin Islands (U.S.) 

# Section 7; Part 105
`unstack()` (Part III)

 - Unstack by multiple levels

In [83]:
world = pd.read_csv('datasets/worldstats.csv', index_col = ["country", "year"])
s = world.stack()
s.head()

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
                  GDP           2.873600e+12
            2013  Population    3.765043e+08
dtype: float64

In [86]:
# Can provide a list of levels to unstack
#  Pull out year then country to columns
s.unstack(level = [1, 0])

year                2015          2014          2013          2012  \
country       Arab World    Arab World    Arab World    Arab World   
Population  3.920223e+08  3.842226e+08  3.765043e+08  3.688026e+08   
GDP         2.530102e+12  2.873600e+12  2.846994e+12  2.773270e+12   

year                2011          2010          2009          2008  \
country       Arab World    Arab World    Arab World    Arab World   
Population  3.610318e+08  3.531122e+08  3.450542e+08  3.368865e+08   
GDP         2.497945e+12  2.103825e+12  1.798878e+12  2.081343e+12   

year                2007          2006      ...               1969  \
country       Arab World    Arab World      ...           Zimbabwe   
Population  3.287666e+08  3.209067e+08      ...       5.036321e+06   
GDP         1.641666e+12  1.404190e+12      ...       1.747999e+09   

year                1968          1967          1966          1965  \
country         Zimbabwe      Zimbabwe      Zimbabwe      Zimbabwe   
Population  4.874113e+06  4.718612e+06  4.568320e+06  4.422132e+06   
GDP         1.479600e+09  1.397002e+09  1.281750e+09  1.311436e+09   

year                1964          1963          1962          1961  \
country         Zimbabwe      Zimbabwe      Zimbabwe      Zimbabwe   
Population  4.279561e+06  4.140804e+06  4.006262e+06  3.876638e+06   
GDP         1.217138e+09  1.159512e+09  1.117602e+09  1.096647e+09   

year                1960  
country         Zimbabwe  
Population  3.752390e+06  
GDP         1.052990e+09  

[2 rows x 11211 columns]

In [87]:
# Order is important
s.unstack(level=[0,1])

country       Arab World                                            \
year                2015          2014          2013          2012   
Population  3.920223e+08  3.842226e+08  3.765043e+08  3.688026e+08   
GDP         2.530102e+12  2.873600e+12  2.846994e+12  2.773270e+12   

country                                                             \
year                2011          2010          2009          2008   
Population  3.610318e+08  3.531122e+08  3.450542e+08  3.368865e+08   
GDP         2.497945e+12  2.103825e+12  1.798878e+12  2.081343e+12   

country                                     ...           Zimbabwe  \
year                2007          2006      ...               1969   
Population  3.287666e+08  3.209067e+08      ...       5.036321e+06   
GDP         1.641666e+12  1.404190e+12      ...       1.747999e+09   

country                                                             \
year                1968          1967          1966          1965   
Population  4.874113e+06  4.718612e+06  4.568320e+06  4.422132e+06   
GDP         1.479600e+09  1.397002e+09  1.281750e+09  1.311436e+09   

country                                                             \
year                1964          1963          1962          1961   
Population  4.279561e+06  4.140804e+06  4.006262e+06  3.876638e+06   
GDP         1.217138e+09  1.159512e+09  1.117602e+09  1.096647e+09   

country                   
year                1960  
Population  3.752390e+06  
GDP         1.052990e+09  

[2 rows x 11211 columns]

In [89]:
# fill_value parameter - Fill in NaN with specific value
#   Notice the Albania values in the 1960s below
s = s.unstack("year", fill_value=0)
s.head()

year                            1960          1961          1962  \
country                                                            
Afghanistan Population  8.994793e+06  9.164945e+06  9.343772e+06   
            GDP         5.377778e+08  5.488889e+08  5.466667e+08   
Albania     Population  0.000000e+00  0.000000e+00  0.000000e+00   
            GDP         0.000000e+00  0.000000e+00  0.000000e+00   
Algeria     Population  1.112489e+07  1.140486e+07  1.169015e+07   

year                            1963          1964          1965  \
country                                                            
Afghanistan Population  9.531555e+06  9.728645e+06  9.935358e+06   
            GDP         7.511112e+08  8.000000e+08  1.006667e+09   
Albania     Population  0.000000e+00  0.000000e+00  0.000000e+00   
            GDP         0.000000e+00  0.000000e+00  0.000000e+00   
Algeria     Population  1.198513e+07  1.229597e+07  1.262695e+07   

year                            1966          1967          1968  \
country                                                            
Afghanistan Population  1.014884e+07  1.036860e+07  1.059979e+07   
            GDP         1.400000e+09  1.673333e+09  1.373333e+09   
Albania     Population  0.000000e+00  0.000000e+00  0.000000e+00   
            GDP         0.000000e+00  0.000000e+00  0.000000e+00   
Algeria     Population  1.298027e+07  1.335420e+07  1.374438e+07   

year                            1969      ...               2006  \
country                                   ...                      
Afghanistan Population  1.084951e+07      ...       2.518362e+07   
            GDP         1.408889e+09      ...       7.057598e+09   
Albania     Population  0.000000e+00      ...       2.992547e+06   
            GDP         0.000000e+00      ...       8.992642e+09   
Algeria     Population  1.414444e+07      ...       3.374933e+07   

year                            2007          2008          2009  \
country                                                            
Afghanistan Population  2.587754e+07  2.652874e+07  2.720729e+07   
            GDP         9.843842e+09  1.019053e+10  1.248694e+10   
Albania     Population  2.970017e+06  2.947314e+06  2.927519e+06   
            GDP         1.070101e+10  1.288135e+10  1.204421e+10   
Algeria     Population  3.426197e+07  3.481106e+07  3.540179e+07   

year                            2010          2011          2012  \
country                                                            
Afghanistan Population  2.796221e+07  2.880917e+07  2.972680e+07   
            GDP         1.593680e+10  1.793024e+10  2.053654e+10   
Albania     Population  2.913021e+06  2.904780e+06  2.900247e+06   
            GDP         1.192695e+10  1.289087e+10  1.231978e+10   
Algeria     Population  3.603616e+07  3.671713e+07  3.743943e+07   

year                            2013          2014          2015  
country                                                           
Afghanistan Population  3.068250e+07  3.162751e+07  3.252656e+07  
            GDP         2.004633e+10  2.005019e+10  1.919944e+10  
Albania     Population  2.896652e+06  2.893654e+06  2.889167e+06  
            GDP         1.278103e+10  1.327796e+10  1.145560e+10  
Algeria     Population  3.818614e+07  3.893433e+07  3.966652e+07  

[5 rows x 56 columns]

# Section 7; Part 106
The `pivot()` method

 - Used to reorganize a dataframe - covert values in a column and convert them to column headers

In [92]:
sales = pd.read_csv('datasets/salesmen.csv', parse_dates = ['Date'])
sales["Salesman"].astype('category')
sales.head()

Date Salesman  Revenue
0 2016-01-01      Bob     7172
1 2016-01-02      Bob     6362
2 2016-01-03      Bob     5982
3 2016-01-04      Bob     7917
4 2016-01-05      Bob     7837

In [93]:
# pivot so that the salesman is a column and renenue is part of that column - Creates a new dataframe
#   index = What to use as the index (unchanged in this case)
#   columns = What columns will be used as the new columns; The more unique values here, the more columns
#   values = What are going to be the values at the index/column intersection
sales.pivot(index="Date", columns="Salesman", values="Revenue")

Salesman     Bob  Dave   Jeb  Oscar  Ronald
Date                                       
2016-01-01  7172  1864  4430   5250    2639
2016-01-02  6362  8278  8026   8661    4951
2016-01-03  5982  4226  5188   7075    2703
2016-01-04  7917  3868  3144   2524    4258
2016-01-05  7837  2287   938   2793    7771
2016-01-06  1744  7859  8702   7794    5930
2016-01-07   918  8597  4250   9728    1933
2016-01-08  9863  3092  9719   5263    5709
2016-01-09  8337  1794  5614   7144    4707
2016-01-10  7543  7105   301   7663    8267
2016-01-11  1053  6851  9489   8888    1340
2016-01-12  4362  7147  8719   3092     279
2016-01-13  6812  6160  2349   6139    7540
2016-01-14  9047  7533  7690   5235    4214
2016-01-15  3594  9883  6917   5507    5346
2016-01-16  9800  3622   474   4947    9636
2016-01-17  7215  4565  8963   3325    6058
2016-01-18  7744  8246  2351   2611    3602
2016-01-19  9560  9229  5512   1905     316
2016-01-20  9863  8358  9509   6005    9352
2016-01-21   547  4229  2453   2798    6781
2016-01-22  3505  4417  2875    879    5166
2016-01-23  6969   358  8853   4245    2479
2016-01-24   696  7817  6705   3798    4835
2016-01-25  4302  6354  2882   1143    9750
2016-01-26  9427  7544  1009    871    8738
2016-01-27  7990  2255  4020   2403    1768
2016-01-28  6271  3419  1735   3691    7264
2016-01-29  5804  1378  6477    418    8545
2016-01-30  3285  6456  2253   7776     658
...          ...   ...   ...    ...     ...
2016-12-02  5570  9842   982   5589    5956
2016-12-03  7910  4284  9363   6009    2252
2016-12-04  2869  8389  5980    537    1484
2016-12-05  4013  3089  8999   3895    2867
2016-12-06  1695  9142  4896   3060    5734
2016-12-07  1010  3707  4634    295    5409
2016-12-08  1149  3073  5930   5038    9951
2016-12-09  4432  6820  8592    258      96
2016-12-10   194  3938  3530   9392    8713
2016-12-11  7676  8805  1134    690    5843
2016-12-12  7290  3917  9004   5621    4524
2016-12-13  7253  9168  8031   3284    5193
2016-12-14  9777  2632  3162   2245    8161
2016-12-15  7933  2051  4408   3449    9557
2016-12-16  6742  1148  3020   1763    2318
2016-12-17  5641  8921  6126   1641    9042
2016-12-18  8445  9203  5843   1022    7979
2016-12-19  3065  4255  8136   3366    2734
2016-12-20  3864  7209  2082    364    5757
2016-12-21  4080  1335  5940   3579     359
2016-12-22  5294  6986  4389   6476     825
2016-12-23  5063  5507     9   6495    9210
2016-12-24  1731  5520  8230   9346    4114
2016-12-25  8325  4004  6569   8321    7490
2016-12-26  6664  3211  1572   6172    6170
2016-12-27  2045  2843  6666    835    2981
2016-12-28   100  8888  1243   3073    6129
2016-12-29  4115  9490  3498   6424    7662
2016-12-30  2577  3594  8858   7088    2570
2016-12-31  3845  6830  9717   8408    2619

[366 rows x 5 columns]

# Section 7; Part 107
`pivot_table()` method

 - Similar to function in Excel
 - Ideal for taking DataFrame and aggregating values as a whole

In [95]:
foods = pd.read_csv('datasets/foods.csv')
foods.head()

First Name  Gender          City Frequency       Item  Spend
0      Wanda  Female      Stamford    Weekly     Burger  15.66
1       Eric    Male      Stamford     Daily    Chalupa  10.56
2    Charles    Male      New York     Never      Sushi  42.14
3       Anna  Female  Philadelphia      Once  Ice Cream  11.01
4    Deborah  Female  Philadelphia     Daily    Chalupa  23.49

In [97]:
# Parameters
#  values = Column we are performing aggfunc on
#  index = Group by...
#  aggfunc = (Default is 'mean')
foods.pivot_table(values="Spend", index="Gender", aggfunc='mean')

# Output shows breakdown of average spend by Gender

Gender
Female    50.709629
Male      49.397623
Name: Spend, dtype: float64

In [98]:
# Total spend by gender
foods.pivot_table(values="Spend", index="Gender", aggfunc='sum')

Gender
Female    25963.33
Male      24106.04
Name: Spend, dtype: float64

In [100]:
# Total sales by item
foods.pivot_table(values="Spend", index="Item", aggfunc='sum')

Item
Burger       7765.73
Burrito      8270.44
Chalupa      7644.52
Donut        8758.76
Ice Cream    8886.99
Sushi        8742.93
Name: Spend, dtype: float64

In [101]:
# Can provide a list to `index` to get further break down
#  Total sales by gender per item
foods.pivot_table(values="Spend", index=["Gender", "Item"], aggfunc='sum')

Gender  Item     
Female  Burger       4094.30
        Burrito      4257.82
        Chalupa      4152.26
        Donut        4743.00
        Ice Cream    4032.87
        Sushi        4683.08
Male    Burger       3671.43
        Burrito      4012.62
        Chalupa      3492.26
        Donut        4015.76
        Ice Cream    4854.12
        Sushi        4059.85
Name: Spend, dtype: float64

In [104]:
# columns parameter
# Want to look at it by Gender/Item and City and create a dataframe based on unique column values in the columns field
foods.pivot_table(values="Spend", index=["Gender", "Item"], columns = "City", aggfunc='sum')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger      1239.04       1639.24   1216.02
       Burrito      978.95       1458.76   1820.11
       Chalupa      876.58       1673.33   1602.35
       Donut       1446.78       1639.26   1656.96
       Ice Cream   1521.62       1479.22   1032.03
       Sushi       1480.29       1742.88   1459.91
Male   Burger      1294.09        938.18   1439.16
       Burrito     1399.40       1312.93   1300.29
       Chalupa     1227.77       1114.23   1150.26
       Donut       1345.27       1249.36   1421.13
       Ice Cream   1603.63       2191.27   1059.22
       Sushi       1396.15       1395.88   1267.82

In [105]:
# Can have a multiindex column too
foods.pivot_table(values="Spend", index=["Gender", "Item"], columns = ["Frequency","City"], aggfunc='sum')

Frequency           Daily                        Monthly               \
City             New York Philadelphia Stamford New York Philadelphia   
Gender Item                                                             
Female Burger      262.67       231.68   144.66   171.86       215.05   
       Burrito     224.45       321.57   195.63   122.74        34.28   
       Chalupa      43.19        23.49    95.70   158.37       289.96   
       Donut       478.10       247.40   124.35   284.53        50.25   
       Ice Cream   262.19       177.69    92.88    92.53        74.51   
       Sushi        81.07       348.53   393.27    93.16        78.71   
Male   Burger      319.46       112.70   197.72   187.29       213.14   
       Burrito     236.21       165.76   276.23   147.54        89.58   
       Chalupa      54.09       274.81   192.64   333.76        90.70   
       Donut       230.00       190.71   129.42   183.73       259.29   
       Ice Cream   178.22       148.69    15.17   152.20       386.55   
       Sushi       225.57       280.93      NaN   430.86       155.59   

Frequency                    Never                          Often   ...     \
City             Stamford New York Philadelphia Stamford New York   ...      
Gender Item                                                         ...      
Female Burger      238.49    97.89       218.97    90.97    94.96   ...      
       Burrito      67.94   189.73       382.30   366.34   103.60   ...      
       Chalupa     161.98    35.15       121.97   156.36    39.73   ...      
       Donut       229.30    56.07       433.58   157.33   130.63   ...      
       Ice Cream   125.85   206.15       156.03    77.66   232.26   ...      
       Sushi       108.39    69.33       190.58    19.56   343.94   ...      
Male   Burger       27.16    90.32        17.31      NaN    55.47   ...      
       Burrito     119.60    86.78       378.32   351.84   189.92   ...      
       Chalupa     343.76   199.09       242.98   138.70   251.52   ...      
       Donut       119.53   219.63        53.65   219.64   139.80   ...      
       Ice Cream    74.36    57.54       321.94   119.77   347.55   ...      
       Sushi        93.03   164.01       245.77   204.90     9.31   ...      

Frequency            Once   Seldom                         Weekly  \
City             Stamford New York Philadelphia Stamford New York   
Gender Item                                                         
Female Burger       95.05    63.16       409.05    97.53   184.35   
       Burrito     168.01   167.54       198.11   234.49    39.69   
       Chalupa      40.59   160.00       274.51   175.25   171.52   
       Donut       158.24    90.81       183.25   104.66   285.56   
       Ice Cream   279.33   242.35       203.10   117.73   227.62   
       Sushi       194.75    87.70       314.24    27.82   410.89   
Male   Burger      326.64   225.68       188.06   159.75    69.69   
       Burrito      30.15   202.40        27.71     9.84   256.74   
       Chalupa        NaN    11.69       130.75   174.02   163.20   
       Donut       139.89    16.25        99.01   163.31    74.44   
       Ice Cream   182.71   270.42       138.44    40.83   223.86   
       Sushi       314.65   160.49        91.26   224.15   150.15   

Frequency                                Yearly                        
City             Philadelphia Stamford New York Philadelphia Stamford  
Gender Item                                                            
Female Burger           32.00   155.02   259.30       123.17   307.03  
       Burrito          62.82   230.91    71.26       116.75   172.98  
       Chalupa          84.41   204.69   157.82       504.44   348.16  
       Donut           208.80   440.06    62.95       175.23   224.48  
       Ice Cream       285.28   125.03   151.67       159.86    15.24  
       Sushi           334.00   315.36   232.41       145.85   362.74  
Male   Burger           99.89   310.21    49.61  

In [106]:
# Other aggregations
#   sum
foods.pivot_table(values="Spend", index=["Gender", "Item"], columns = "City", aggfunc='sum')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger      1239.04       1639.24   1216.02
       Burrito      978.95       1458.76   1820.11
       Chalupa      876.58       1673.33   1602.35
       Donut       1446.78       1639.26   1656.96
       Ice Cream   1521.62       1479.22   1032.03
       Sushi       1480.29       1742.88   1459.91
Male   Burger      1294.09        938.18   1439.16
       Burrito     1399.40       1312.93   1300.29
       Chalupa     1227.77       1114.23   1150.26
       Donut       1345.27       1249.36   1421.13
       Ice Cream   1603.63       2191.27   1059.22
       Sushi       1396.15       1395.88   1267.82

In [107]:
# mean
foods.pivot_table(values="Spend", index=["Gender", "Item"], columns = "City", aggfunc='mean')

City               New York  Philadelphia   Stamford
Gender Item                                         
Female Burger     51.626667     52.878710  45.037778
       Burrito    42.563043     52.098571  53.532647
       Chalupa    46.135789     52.291563  64.094000
       Donut      46.670323     54.642000  48.734118
       Ice Cream  56.356296     46.225625  46.910455
       Sushi      47.751290     58.096000  45.622187
Male   Burger     58.822273     44.675238  46.424516
       Burrito    55.976000     43.764333  46.438929
       Chalupa    49.110800     48.444783  50.011304
       Donut      44.842333     37.859394  49.004483
       Ice Cream  55.297586     53.445610  42.368800
       Sushi      51.709259     49.852857  70.434444

In [108]:
# count
foods.pivot_table(values="Spend", index=["Gender", "Item"], columns = "City", aggfunc='count')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger           24            31        27
       Burrito          23            28        34
       Chalupa          19            32        25
       Donut            31            30        34
       Ice Cream        27            32        22
       Sushi            31            30        32
Male   Burger           22            21        31
       Burrito          25            30        28
       Chalupa          25            23        23
       Donut            30            33        29
       Ice Cream        29            41        25
       Sushi            27            28        18

In [109]:
# max
foods.pivot_table(values="Spend", index=["Gender", "Item"], columns = "City", aggfunc='max')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger        98.96         97.79     85.06
       Burrito       92.25         96.79     99.21
       Chalupa       98.43         99.29     98.78
       Donut         95.63         96.52     91.75
       Ice Cream     97.83         88.14     97.44
       Sushi         99.51         99.02     95.43
Male   Burger        90.32         99.68     97.20
       Burrito       98.04         93.27     95.07
       Chalupa       96.44         98.40     99.87
       Donut         86.70         93.12     99.26
       Ice Cream     97.65         99.24     99.17
       Sushi         93.85         97.12     98.48

In [110]:
# min
foods.pivot_table(values="Spend", index=["Gender", "Item"], columns = "City", aggfunc='min')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger         2.25          1.97      6.24
       Burrito        1.02          1.04      1.18
       Chalupa        1.96          9.35      9.09
       Donut          3.15          2.13      1.68
       Ice Cream     13.39          7.61      8.80
       Sushi          2.52         11.68      8.20
Male   Burger         5.43          1.71      2.83
       Burrito       15.90          8.58      3.64
       Chalupa       11.61          1.94     10.56
       Donut          1.49          1.26      6.63
       Ice Cream     14.06          4.89      3.43
       Sushi          3.28          2.01     32.15

In [112]:
# pivot_table is available directly on the pandas object too (not just a dataframe)
# Requires you pass a dataframe as the first parameter
pd.pivot_table(data=foods, values="Spend", index=["Gender", "Item"], columns = "City", aggfunc='min')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger         2.25          1.97      6.24
       Burrito        1.02          1.04      1.18
       Chalupa        1.96          9.35      9.09
       Donut          3.15          2.13      1.68
       Ice Cream     13.39          7.61      8.80
       Sushi          2.52         11.68      8.20
Male   Burger         5.43          1.71      2.83
       Burrito       15.90          8.58      3.64
       Chalupa       11.61          1.94     10.56
       Donut          1.49          1.26      6.63
       Ice Cream     14.06          4.89      3.43
       Sushi          3.28          2.01     32.15

# Section 7; Part 108
`pd.melt()` method

 - Works as the reverse of a pivot_table

In [113]:
sales = pd.read_csv('datasets/quarters.csv')
sales

Salesman      Q1      Q2      Q3      Q4
0    Boris  602908  233879  354479   32704
1      Bob   43790  514863  297151  544493
2    Tommy  392668  113579  430882  247231
3   Travis  834663  266785  749238  570524
4   Donald  580935  411379  110390  651572
5      Ted  656644   70803  375948  321388
6      Jeb  486141  600753  742716  404995
7    Stacy  479662  742806  770712    2501
8   Morgan  992673  879183   37945  293710

Goal is to unpivot the above so it is in a tabular format

In [116]:
# melt works on the pandas object
#   frame = Dataframe we are operating on
#   id_vars = Column that will be preserved - in this case, we want to keep the salesmen, the Q1-Q4 will be moved into table
# By default, the new columns will be "variable" and "value"
#   var_name = Name of new variable column
#   value_name = Name of new value column
pd.melt(frame=sales, id_vars="Salesman", var_name="Quarter", value_name="Revenue")

Salesman Quarter  Revenue
0     Boris      Q1   602908
1       Bob      Q1    43790
2     Tommy      Q1   392668
3    Travis      Q1   834663
4    Donald      Q1   580935
5       Ted      Q1   656644
6       Jeb      Q1   486141
7     Stacy      Q1   479662
8    Morgan      Q1   992673
9     Boris      Q2   233879
10      Bob      Q2   514863
11    Tommy      Q2   113579
12   Travis      Q2   266785
13   Donald      Q2   411379
14      Ted      Q2    70803
15      Jeb      Q2   600753
16    Stacy      Q2   742806
17   Morgan      Q2   879183
18    Boris      Q3   354479
19      Bob      Q3   297151
20    Tommy      Q3   430882
21   Travis      Q3   749238
22   Donald      Q3   110390
23      Ted      Q3   375948
24      Jeb      Q3   742716
25    Stacy      Q3   770712
26   Morgan      Q3    37945
27    Boris      Q4    32704
28      Bob      Q4   544493
29    Tommy      Q4   247231
30   Travis      Q4   570524
31   Donald      Q4   651572
32      Ted      Q4   321388
33      Jeb      Q4   404995
34    Stacy      Q4     2501
35   Morgan      Q4   293710